File: backend/src/main/java/com/mealapp/service/AuthService.java
package com.mealapp.service;

import com.mealapp.controller.AuthResponse;
import com.mealapp.model.User;
import com.mealapp.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {
    
    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private PasswordEncoder passwordEncoder;
    
    @Autowired
    private JwtService jwtService;
    
    // User login
    public AuthResponse login(String email, String password) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        if (!passwordEncoder.matches(password, user.getPassword())) {
            throw new RuntimeException("Invalid password");
        }
        
        String token = jwtService.generateToken(user);
        return new AuthResponse(token, user);
    }
    
    // User registration
    public AuthResponse register(String name, String email, String password) {
        if (userRepository.findByEmail(email).isPresent()) {
            throw new RuntimeException("Email already exists");
        }
        
        User user = new User();
        user.setName(name);
        user.setEmail(email);
        user.setPassword(passwordEncoder.encode(password));
        
        User savedUser = userRepository.save(user);
        String token = jwtService.generateToken(savedUser);
        
        return new AuthResponse(token, savedUser);
    }
    
    // Refresh token
    public AuthResponse refreshToken(String token) {
        String email = jwtService.extractEmail(token.replace("Bearer ", ""));
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
        
        String newToken = jwtService.generateToken(user);
        return new AuthResponse(newToken, user);
    }
    
    // Get current user
    public User getCurrentUser(String token) {
        String email = jwtService.extractEmail(token.replace("Bearer ", ""));
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }
    
    // Update user profile
    public User updateProfile(String token, User userDetails) {
        User user = getCurrentUser(token);
        
        user.setName(userDetails.getName());
        // Don't update email and password here
        
        return userRepository.save(user);
    }
    
    // Change password
    public void changePassword(String token, String currentPassword, String newPassword) {
        User user = getCurrentUser(token);
        
        if (!passwordEncoder.matches(currentPassword, user.getPassword())) {
            throw new RuntimeException("Current password is incorrect");
        }
        
        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
    }
}

File: frontend/src/services/authService.js
import api from './api'

export const authService = {
  // Login user
  login: async (email, password) => {
    const response = await api.post('/auth/login', { email, password })
    const { token, id, name, email: userEmail, role } = response.data
    
    const user = { id, name, email: userEmail, role }
    
    localStorage.setItem('authToken', token)
    localStorage.setItem('user', JSON.stringify(user))
    
    return user
  },

  // Register new user
  register: async (userData) => {
    const response = await api.post('/auth/register', userData)
    return response.data
  },

  // Logout user
  logout: () => {
    localStorage.removeItem('authToken')
    localStorage.removeItem('user')
    window.location.href = '/login'
  },

  // Get current user
  getCurrentUser: () => {
    const userStr = localStorage.getItem('user')
    return userStr ? JSON.parse(userStr) : null
  },

  // Check if user is authenticated
  isAuthenticated: () => {
    return !!localStorage.getItem('authToken')
  },

  // Get auth token
  getToken: () => {
    return localStorage.getItem('authToken')
  },

  // Refresh token
  refreshToken: async () => {
    const response = await api.post('/auth/refresh')
    const { token } = response.data
    localStorage.setItem('authToken', token)
    return token
  },

  // Update user profile
  updateProfile: async (userData) => {
    const response = await api.put('/auth/profile', userData)
    const { user } = response.data
    localStorage.setItem('user', JSON.stringify(user))
    return response.data
  },

  // Change password
  changePassword: async (currentPassword, newPassword) => {
    const response = await api.put('/auth/change-password', {
      currentPassword,
      newPassword
    })
    return response.data
  }
}